plasmapy/diagnostics/charged_particle_radiography/synthetic_radiography.py

```diff
@@ -897,107 +771,6 @@ def _remove_deflected_particles(self):
-            # Note that this interpolation step is BY FAR the slowest part of the push
-            # loop. Any speed improvements will have to come from here.
-            if self.field_weighting == "volume averaged":
-                _Ex, _Ey, _Ez, _Bx, _By, _Bz = grid.volume_averaged_interpolator(
-                    pos,
-                    "E_x",
-                    "E_y",
-                    "E_z",
-                    "B_x",
-                    "B_y",
-                    "B_z",
-                    persistent=True,
-                )
-            elif self.field_weighting == "nearest neighbor":
-                _Ex, _Ey, _Ez, _Bx, _By, _Bz = grid.nearest_neighbor_interpolator(
-                    pos,
-                    "E_x",
-                    "E_y",
-                    "E_z",
-                    "B_x",
-                    "B_y",
-                    "B_z",
-                    persistent=True,
-                )
-
-            # Interpret any NaN values (points off the grid) as zero
-            # Do this before adding to the totals, because 0 + nan = nan
-            _Ex = np.nan_to_num(_Ex, nan=0.0 * u.V / u.m)
-            _Ey = np.nan_to_num(_Ey, nan=0.0 * u.V / u.m)
-            _Ez = np.nan_to_num(_Ez, nan=0.0 * u.V / u.m)
-            _Bx = np.nan_to_num(_Bx, nan=0.0 * u.T)
-            _By = np.nan_to_num(_By, nan=0.0 * u.T)
-            _Bz = np.nan_to_num(_Bz, nan=0.0 * u.T)
-
-            # Add the values interpolated for this grid to the totals
-            Ex += _Ex
-            Ey += _Ey
-            Ez += _Ez
-            Bx += _Bx
-            By += _By
-            Bz += _Bz
-
-        # Create arrays of E and B as required by push algorithm
-        E = np.array(
-            [Ex.to(u.V / u.m).value, Ey.to(u.V / u.m).value, Ez.to(u.V / u.m).value]
-        )
-        E = np.moveaxis(E, 0, -1)
-        B = np.array([Bx.to(u.T).value, By.to(u.T).value, Bz.to(u.T).value])
-        B = np.moveaxis(B, 0, -1)
-
-        # Calculate the adaptive timestep from the fields currently experienced
-        # by the particles
-        # If user sets dt explicitly, that's handled in _adaptive_dt
-        dt = self._adaptive_dt(Ex, Ey, Ez, Bx, By, Bz)
-
-        # TODO: Test v/c and implement relativistic Boris push when required
-        # vc = np.max(v)/_c
-
-        # If dt is not a scalar, make sure it can be multiplied by an
-        # [nparticles, 3] shape field array
-        if dt.size > 1:
-            dt = dt[:, np.newaxis]
-
-        x = self.x[self.grid_ind, :]
-        v = self.v[self.grid_ind, :]
-        boris_push(x, v, B, E, self.q, self.m, dt)
-        self.x[self.grid_ind, :] = x
-        self.v[self.grid_ind, :] = v
-
-    @property
-    def on_any_grid(self):
-        """
-        Binary array for each particle indicating whether it is currently
-        on ANY grid.
-        """
-        return np.sum(self.on_grid, axis=-1) > 0
-
     def _stop_condition(self):
```

Comment:
Let's have this be a function (could be a static method of this class) that gets passed to ParticleTracker 

e.g. 

```python
class SyntheticRadiograh: 

@staticmethod
def _stop_condition(particle_tracker):

# Determine how many particles have entered and then left the grid 

if < threshold: 
return fract_entered # for progress bar 

else:
return True


class ParticleTracker:

def run( stop_condition=None):

# If None, use a default stopping condition function

stop=False
while not stop:
stop = _stopping_condition(self)
if stop is not True: 
nentered = stop
stop = False



```

Reply:
That's weird - maybe it doesn't like that it takes an instance of the class as an argument? 

What if each stopping condition was just a method (e.g. `_time_elapsed_stopping_condition`, `_particles_on_grid_stopping_condition` and then `self._stopping_condition` was an attribute that you could set equal to one of the other methods, choosing which one is used?